Add provider_shopper

filiph opened the pull request:
This is the code for the sample at https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple (when that's rewritten to use `provider` instead of `scoped_model`).

filiph commented:
@RedBrogdon I'm not sure if this belongs here, so feel free to just tell me to leave it in my fork. It's this app:

![shopper app gif](https://flutter.dev/assets/development/data-and-backend/state-mgmt/model-shopper-screencast-e0ada0e83cd8e7fdcad84167b8f7ffd7eb5ef85b0cb8957f03c6f05bd16b1cea.gif)

Anas1981 commented:
Hi Dear ,
In-fact I'm looking for Job in website design or  assist on any app initiative if you want some one to help you please feel free to contact me.
Anas Altahir

RedBrogdon commented:
(hopefully you don't mind a bunch of questions 😄)

Is this mostly a thing for the website, and you just need a place to store it?
How does this related to provider_counter? Would it supercede that sample? Do we really need both?
Are you committed to maintaining this and keeping it up to date, or is it more of a one-off?

RedBrogdon commented:
@filiph gentle ping. 😄

filiph commented:
Sorry for the late reply:

> Is this mostly a thing for the website, and you just need a place to store it?

Place to store and test this. And since it's one of the few samples that's also explained on the official site, I thought it would be fitting to have here.

> How does this related to provider_counter? Would it supercede that sample? Do we really need both?

provider_counter serves two purposes: 

1. Is almost the simplest possible app that uses Provider. The hello world, if you will.
2. It's a nice counterpart to the counter app using setState that we all know and love. People have been traditionally confused about that ("is setState the right way to manage apps?")

provider_shopper is a more complete  and realistic sample app, with multiple routes, a `ListView`, and a more realistic `ChangeNotifier`

> Are you committed to maintaining this and keeping it up to date, or is it more of a one-off?

I'm committed. If this isn't covered, then we're doing a bad job showing people how to do state management, imho.

RedBrogdon commented on provider_shopper/lib/screens/cart.dart:
```diff
@@ -0,0 +1,74 @@
+class _CartList extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    var textTheme = Theme.of(context).textTheme.title;
```
nit: `textStyle` would be a more accurate name here than `textTheme`.

RedBrogdon commented on provider_shopper/lib/screens/cart.dart:
```diff
@@ -0,0 +1,74 @@
+class _CartTotal extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    var textTheme = Theme.of(context).textTheme.display4.copyWith(fontSize: 48);
```
`textStyle`

RedBrogdon commented on provider_shopper/lib/screens/cart.dart:
```diff
@@ -0,0 +1,74 @@
+    var textTheme = Theme.of(context).textTheme.title;
+    var cart = Provider.of<CartModel>(context);
+
+    return ListView(
```
Partially at @xster's urging, I've started using ListView.builder everywhere possible, since it's more efficient. You're never going to have a zillion items in this list, but it'd still be a good thing to show.

Also (and this is a bit of a nit), consider using a ListTile ratehr than just a Text and sticking the bullet character in the `leading`  property.

RedBrogdon commented on provider_shopper/lib/screens/catalog.dart:
```diff
@@ -0,0 +1,89 @@
+          SliverToBoxAdapter(child: SizedBox(height: 12)),
+          SliverList(
+              delegate:
+                  SliverChildBuilderDelegate((_, index) => _MyListItem(index))),
```
Given that this is sample code, readers might benefit more from seeing the name of the first parameter here.

RedBrogdon commented on provider_shopper/lib/screens/catalog.dart:
```diff
@@ -0,0 +1,89 @@
+    var cart = Provider.of<CartModel>(context);
+
+    return FlatButton(
+      onPressed: cart.items.contains(item) ? null : () => cart.add(item),
```
I don't think I've ever used a ternary expression with a function like this. It's a nice touch to disable the button!

RedBrogdon commented on provider_shopper/lib/screens/catalog.dart:
```diff
@@ -0,0 +1,89 @@
+      actions: [
+        IconButton(
+            icon: Icon(Icons.shopping_cart),
+            onPressed: () => Navigator.pushNamed(context, '/cart')),
```
nit: throw a comma after the close of pushNamed so this will indent a little better.

RedBrogdon commented on provider_shopper/pubspec.yaml:
```diff
@@ -0,0 +1,29 @@
+name: provider_shopper
+description: A Flutter app that uses Provider.
```
nit: Update this to something more descriptive.

RedBrogdon commented on provider_shopper/pubspec.yaml:
```diff
@@ -0,0 +1,29 @@
+version: 1.0.0+1
+
+environment:
+  sdk: ">=2.2.2 <3.0.0"
```
`>=2.3.0-dev` unless you have a reason for keeping it lower.

RedBrogdon commented on provider_shopper/pubspec.yaml:
```diff
@@ -0,0 +1,29 @@
+    - family: Corben
+      fonts:
+        - asset: fonts/Corben/Corben-Bold.ttf
+          weight: 700
```
nit: newline.

xster commented:
What's the login screen meant to teach? Can you log out and affect UI downstream?

filiph commented:
Hi @RedBrogdon, I've addressed your feedback from before and updates the sample in other ways that we discussed this week. Some of these changes are significant. Can you please re-review?

@xster: I got rid of the Login screen altogether. It was supposed to teach use of multiple Providers, but I found another way to do it.

RedBrogdon commented on provider_shopper/lib/models/cart.dart:
```diff
@@ -0,0 +1,29 @@
+  /// Internal, private state of the cart. Stores the ids of each item.
+  final List<int> _itemIds;
+
+  CartModel(this._catalog, CartModel previous)
```
nit: Throw a comment on here to explain when this would be used (also, it's the only method in here with no comment).

RedBrogdon commented on provider_shopper/lib/models/catalog.dart:
```diff
@@ -0,0 +1,54 @@
+  ];
+
+  /// Get item by [id].
+  Item getById(int id) => Item(id, _itemNames[id % _itemNames.length]);
```
The modulo operation in here was confusing at first glance. Now that I've seen the rest of the code, I understand that you're doing it so that you have an infinitely scrolling list. That's a cool piece of UI to show, but I'm concerned that people won't be able to understand this file until they go further into the app.

Consider:

* Removing the infinite list and just showing the items in the catalog once.
* Commenting here and in the colors file to explain that you're faking an infinite list of products by repeating names and colors, and that's why the modulo operation is there (instead of a thrown IndexOutOfRange exception or something).

RedBrogdon commented on provider_shopper/lib/screens/cart.dart:
```diff
@@ -0,0 +1,79 @@
+                child: _CartList(),
+              ),
+            ),
+            Container(height: 4, color: Colors.black),
```
The Divider widget might be more idiomatic than a Container here.

RedBrogdon commented on provider_shopper/lib/screens/cart.dart:
```diff
@@ -0,0 +1,79 @@
+class _CartTotal extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    var textStyle = Theme.of(context).textTheme.display4.copyWith(fontSize: 48);
```
nit: consider a more descriptive name for this ("bigTitleStyle") or something.

RedBrogdon commented on provider_shopper/lib/screens/catalog.dart:
```diff
@@ -0,0 +1,91 @@
+class _MyListItem extends StatelessWidget {
+  final int index;
+
+  _MyListItem(this.index) : super(key: ValueKey(index));
```
This isn't a method to create the key that I've seen elsewhere. Are there other widgets that do it? I'm used to the containing widget being responsible for creating and assigning the key.

filiph replied:
You're right. This was a remnant of some perf experiment. Got rid of it.

RedBrogdon commented on provider_shopper/lib/screens/catalog.dart:
```diff
@@ -0,0 +1,91 @@
+              ),
+            ),
+            SizedBox(width: 24),
+            Text(item.name, style: textTheme),
```
Unlikely to matter for your app, but I believe that unless the `Text` is wrapped in a `Flexible` (like `Expanded`), it will never break a line and could overflow to the right. Consider removing the `Spacer` and wrapping `Text` in an `Expanded` instead.

filiph replied:
Damn, you're right, of course. I was so enamored with Spacer (a new find for me) that I _removed_ Expanded and put a Spacer here, thinking how clever I was. 🤦‍♂

RedBrogdon commented on travis_script.sh:
```diff
@@ -8,6 +8,7 @@ declare -a  PROJECT_NAMES=(
     "platform_design"
     "platform_view_swift" \
     "provider_counter" \
+    "provider_shopper" \
```
Thanks for remembering to add it here. Sine this PR was started, we also began standardizing analysis_options.yaml files as well. Can you copy in the file from one of the other samples?

filiph commented on provider_shopper/lib/models/catalog.dart:
```diff
@@ -0,0 +1,54 @@
+  /// Get item by its position in the catalog.
+  Item getByPosition(int index) {
+    // In our simplified case, an item's position in the catalog is also its id.
+    return getById(index);
```
I've thought about it, but I think rewriting would make the matter even worse than it is. We would still need to explain why we're using the same code as above, otherwise it looks like a copy-paste bug.

```dart
  /// Get item by [id].
  Item getById(int id) => Item(id, _itemNames[id % _itemNames.length]);

  /// Get item by [position].
  Item getByPosition(int position) => Item(position, _itemNames[position % _itemNames.length]);
```

 And if we explain, it's better to just call the other method. I don't feel strongly about this, though, so feel free to push back.

filiph commented:
Thanks for the review, @RedBrogdon! Ready for another go.

RedBrogdon commented:
The only thing it needs now is copyright headers on the Dart files. Once those are in place, LGTM!
